### PR TITLE
React Progressive Web App

### PR DESCRIPTION
**Pull request by ReactifyStudio**
### Article Checklist
This was an article about creating a React Todo PWA
#### Preliminary Checks - Formatting and Structure 
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**
- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**
- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
- [x] In your PR - add a comment that states **["This fixes #(enter your Idea Content Form #)"](https://youtu.be/TKJ4RdhyB5Y?t=278)** to link and close your Idea Content (issue) when the PR is merged.

**NOTE: (1st time contributors - submit ONLY one article to get familiar with the process and style-once published, you can then submit any number of articles.)**
      
#### Grammar Checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) -  [Grammarly](https://grammarly.com) 
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical Checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**

### Contribution Guidelines
For first-time contributors and for more details, **see [Contributing Guidelines](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md)**

### CODE REVIEW

**Comment by ReactifyStudio** [comment]
This fixes #895

**Comment by zolomohan** [comment]
Please update PWA to Progressive Web Application in your PR Title.

**Comment by hectorkambow** [comment]
plagiarism check done

**Review by zolomohan — CHANGES_REQUESTED** [review]
Hi @ReactifyStudio.

Upon reviewing this article, I noticed that the Todo application you have created is architecturally not optimized. You are passing all the data to low-level components and performing operations from the low-level components instead of performing it from parent containers.

- The `Delete Todo` and the `Complete Task` functions should be performed from the `TodoList` component rather than the `Todo` component. The Todo component should not have access to the `setTodos` method nor should it know about all the `todos` in the list. You should write a function to perform `delete` and `complete` and pass it down to the Todo component.

- The state for the Form input should not reside in the `App` component. It should be inside the `Form` component. You're are passing the `todos` and the `setTodos` to the Form component which is unnecessary. Instead, you should write a function to add the todo to the `todos state` and pass only that function to the `Form` component and call it when the Add button is clicked.

Also, run the article through Grammarly and Hemmingway once. I noticed a lot of grammatical errors. Some sentences should be written more precisely.

**Comment by ReactifyStudio** [comment]
I have by now refactored the entire code and done grammar check up

**Comment by ReactifyStudio** [comment]
@zolomohan it's done. Have completed the changes and improved grammar.

File: articles/react-pwa/index.md [hunk 1]
```diff
@@ -0,0 +1,508 @@
...
+        setInputText(e.target.value);
+    }
+    const submitTodoHandler=(e)=>{
+        e.preventDefault();
+        if(!inputText){
+            return
+        }
+        addTodo({text:inputText,completed:false,id:uuidv4()});
+        setInputText("");
+    }
```
Comment (zolomohan) [inline articles/react-pwa/index.md:89]: You can change this too

```JavaScript
if(inputText){
  addTodo({text:inputText, completed:false, id:uuidv4()});
}

```

File: articles/react-pwa/index.md [hunk 2]
```diff
@@ -0,0 +1,508 @@
...
+```javascript
+import React,{useState} from 'react';
+
+import Form from './components/Form';
+
+const App=()=>{
+
+  const addTodo= todo=>{
+          const newTodos=[todo,...todos];
+           setTodos(newTodos);
```
Comment (zolomohan) [inline articles/react-pwa/index.md:119]: setTodos is not declared yet.

File: articles/react-pwa/index.md [hunk 3]
```diff
@@ -0,0 +1,508 @@
...
+  return (
+    <div>
+      <h1>TODO PWA</h1>
+      <Form addTodo={addTodo}/>
+    </div>
+  );
+}
+
+export default App;
+```
```
Comment (zolomohan) [inline articles/react-pwa/index.md:131]: I think you can remove this snippet, move to the TodoList component, and Introduce App.js in that section.

File: articles/react-pwa/index.md [hunk 4]
```diff
@@ -0,0 +1,508 @@
...
+
+The todo will be an object with a structure like:
+```javascript
+    {
+      id:v4(),
+      task,
+      complete:false
+    }
+```
```
Comment (zolomohan) [inline articles/react-pwa/index.md:179]: You should explain this in the Form component since you've already introduced it there.

**Review by zolomohan — CHANGES_REQUESTED** [review]
@ReactifyStudio.
You've done a great job refactoring the code. 
I've left some minor suggestions. Ping me when you have implemented them.

We usually do a Preliminary Revision before moving to the Tech Review. But in this case, I knew there's going to be a lot of refactoring, so I didn't start the preliminary revision yet. 

Ping me when you have implemented these corrections. I'll do a prelim revision on the grammar and flow of the article. Later, I'll conduct a Tech review.

File: articles/react-pwa/index.md (hunk 3 above)
Comment (ReactifyStudio) [reply articles/react-pwa/index.md:131]: You mean re implementation by code refactoring, or the explanation workflow?

**Comment by ReactifyStudio** [comment]
@zolomohan  please check my changes.

**Comment by zolomohan** [comment]
Hi @ReactifyStudio 

Somehow, the article got messed up. The article is copied twice in the same .MD file. Can you please fix it?

Also, I noticed that for each section of your article, You're providing the whole code snippet and explaining it in a single paragraph. I think it would be better if you break the snippets and explain them separately.

For example, In the Form component, You gave the entire file at once, and then you are explaining hooks, useState, the form, all in one paragraph. This will be hard for the readers to keep track of. 

Instead, you can break the code snippet, explain it step by step, and in the end, you can provide the entire code snippet for reference.

**Comment by ReactifyStudio** [comment]
@zolomohan  is everything okay? Please check it out.

**Comment by ReactifyStudio** [comment]
@zolomohan  check it out. Changes done.

File: articles/react-pwa/index.md [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+export default TodoList;
+```
+
+In this component:
+
+- The `TodoList` is receiving `todos` as an array of objects which are accessed by destructuring its props.
+
+- In the return of our function, we map through the `todos` and return a JSX syntax with a new Todo component as the todo item.
+
+- The `removeTodo` passes an `id` that uniquely identifies the `todo` in our list. Whatever todo that does not match the provided `id` is filtered and stored in the state using `setTodos`.
```
Comment (zolomohan) [inline articles/react-pwa/index.md:196]: `The removeTodo passes an id`

You should say that the `removeTodo` is a function and it accepts an ID.

File: articles/react-pwa/index.md [hunk 6]
```diff
@@ -0,0 +1,539 @@
...
+### The React Application so far
+Your app so far should look like this:
+
+![React-PWA-image](/engineering-education/react-pwa/react-pwa1.png)
+
+![React-PWA-image2](/engineering-education/react-pwa/react-pwa2.png)
+
+
+### Adding functionality to make it a Progressive Web App
+In this part, we will now make our application to implement the Progressive Web App features. Some features of a progressive web app include:
```
Comment (zolomohan) [inline articles/react-pwa/index.md:340]: You can just say, 

`In this part, we will implement the Progressive Web App features. Some features of a progressive web app include:`

File: articles/react-pwa/index.md [hunk 7]
```diff
@@ -0,0 +1,539 @@
...
+
+1. Safety as it should rely on the HTTPS protocol.
+
+2. Must be discoverable and installable.
+
+3. A Progressive web App must be able to work offline.
```
Comment (zolomohan) [inline articles/react-pwa/index.md:346]: This should be the first point.

**Review by zolomohan — CHANGES_REQUESTED** [review]
Hi @ReactifyStudio

Since this is a React App, it will be better if you start explaining from the App component.

You can introduce the Form and the TodoList component, and say that we will create them in the next steps.

Then, you can start with the Form component and the TodoList component.

**Comment by ReactifyStudio** [comment]
Hi @zolomohan , thanks for your review, please checkout the updates.

File: articles/react-pwa/index.md [hunk 8]
```diff
@@ -0,0 +1,526 @@
...
+          .catch(err=>console.log("Error",err))
+        })
+      }
+    </script>
+  </body>
+</html>
+```
+
+### Performance of the App
+I used Lighthouse to generate an audit report. Our App scores 89/100 on performance.
```
Comment (zolomohan) [inline articles/react-pwa/index.md:509]: Add a Link to Lighthouse. Briefly explain what it is.

**Review by zolomohan — CHANGES_REQUESTED** [review]
Isn't the app supposed to be fully functional offline since this app doesn't have any online features?

If I go offline, all I see is this. I installed this from the Netlify link that you provided. If I install from the localhost, I'm not seeing anything if I kill the server.

image

**Comment by ReactifyStudio** [comment]
@zolomohan , I can see the issues, since I am using CRA version 4, I will re-implement most of the PWA part. Thanks.

**Comment by zolomohan** [comment]
Ping me when you fix it.

**Comment by ReactifyStudio** [comment]
Implementation in the index HTML file can only fetch the static assets

**Comment by ReactifyStudio** [comment]
@zolomohan , check my changes . I operate from Kenyan time zone. Have used Workbox since the App was just about adding functionality briefly on the PWA part.

File: articles/react-pwa/index.md [hunk 9]
```diff
@@ -0,0 +1,465 @@
...
+service workers, and also installing it on the end user's device home screen.
+
+### Registering a service worker
+A service worker is a script that runs in the browser's background. It handles network intercepts from requests and managing caching for offline availability. The create-react-app provided all the tools needed to create a Progressive Web App in React. To configure this to run offline, we need to register the generated service worker file. In the project, open the file named `index.js` and find the code that looks like:
+```javascript
+// If you want your app to work offline and load faster, you can change
+// unregister() to register() below. Note this comes with some pitfalls.
+// Learn more about service workers: https://cra.link/PWA
+serviceWorkerRegistration.unregister() 
+```
```
Comment (zolomohan) [inline articles/react-pwa/index.md:353]: Hi, I created a new react app using `npx create-react-app` and I'm not finding this piece of code in the `index.js` file.

Comment (zolomohan) [reply articles/react-pwa/index.md:353]: I don't find the `serviceWorker.js` file too.

Comment (ReactifyStudio) [reply articles/react-pwa/index.md:353]: @zolomohan, I had mentioned that one should use `npx create-react-app react-pwa --template cra-template-pwa`. The --template cra-template-pwa will add it

**Comment by ReactifyStudio** [comment]
Use the --template cra-template-pwa option in the npx create-react-app
command . It works for me

On Jan 9, 2021 18:46, "Mohan Raj" <notifications@github.com> wrote:

*@zolomohan* commented on this pull request.
------------------------------

In articles/react-pwa/index.md
<https://github.com/section-io/engineering-education/pull/1068#discussion_r554440749>
:

> +```javascript
+// If you want your app to work offline and load faster, you can change
+// unregister() to register() below. Note this comes with some pitfalls.
+// Learn more about service workers: https://cra.link/PWA
+serviceWorkerRegistration.unregister()
+```

I don't find the serviceWorker.js file too.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/section-io/engineering-education/pull/1068#discussion_r554440749>,
or unsubscribe

File: articles/react-pwa/index.md (hunk 9 above)
Comment (ReactifyStudio) [reply articles/react-pwa/index.md:353]: The files will be service-worker.js and serviceWorkerRegistration.js

Comment (ReactifyStudio) [reply articles/react-pwa/index.md:353]: Being my first article, I apologise for all inconveniences I've caused. I will be improving on the future tutorials and writing topics that am well polished 👍.

Comment (zolomohan) [reply articles/react-pwa/index.md:353]: I didn't notice that you have updated the create command. My bad. Thanks for clearing it up.

**Comment by ReactifyStudio** [comment]
Okay, check it out and if any issues I'll correct them

**Review by zolomohan — APPROVED** [review]

**Comment by hectorkambow** [comment]
@ReactifyStudio  Great work - please be sure to take our peer reviewer comments and edits and apply them to future articles when applicable. 
🚀  
cc @zolomohan

**Pull request merged**